File: app/src/main/java/com/mobile/activities/ActivateMoviePassCard.java
package com.mobile.activities;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.DecelerateInterpolator;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.mobile.Constants;
import com.mobile.home.HomeActivity;
import com.mobile.model.Screening;
import com.mobile.network.RestClient;
import com.mobile.requests.CardActivationRequest;
import com.mobile.responses.CardActivationResponse;
import com.moviepass.R;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import io.card.payment.CardIOActivity;
import io.card.payment.CreditCard;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ActivateMoviePassCard extends AppCompatActivity {

    View progress;
    TextView activateInstructions, activateManualInput, activateSubmitButton;
    EditText activateDigits;
    ImageView activateScanCardIcon, activateXOut;
    String digits;
    Screening screeningObject;
    String selectedShowTime;
    private final static int REQUEST_CAMERA_CODE = 0;
    private static String CAMERA_PERMISSIONS[] = new String[]{
            Manifest.permission.CAMERA
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.ac_activate_movie_pass_card);


        activateInstructions = findViewById(R.id.ACTIVATECARD_INTRUCTIONS);
        activateManualInput = findViewById(R.id.ACTIVATECARD_MANULINPUT);
        activateDigits = findViewById(R.id.ACTIVATE_DIGITS);
        activateSubmitButton = findViewById(R.id.ACTIVATE_BUTTON);
        activateScanCardIcon = findViewById(R.id.ACTIVATECARD_SCAN_ICON);
        activateXOut = findViewById(R.id.ACTIVATECARD_X_OUT);
        progress = findViewById(R.id.progress);
        activateScanCardIcon.setOnClickListener(v -> scanCard());

        Intent intent = getIntent();
        if (getIntent() != null) {
            screeningObject = intent.getParcelableExtra(Constants.SCREENING);
            selectedShowTime = getIntent().getStringExtra(Constants.SHOWTIME);
        }


        activateXOut.setOnClickListener(v -> {
            Intent closeIntent = new Intent(ActivateMoviePassCard.this, HomeActivity.class);
            startActivity(closeIntent);
        });

        activateManualInput.setOnClickListener(v -> {
            fadeOut(activateScanCardIcon);
            activateScanCardIcon.setVisibility(View.GONE);
            fadeOut(activateManualInput);
            activateManualInput.setVisibility(View.GONE);


            fadeIn(activateSubmitButton);
            activateSubmitButton.setVisibility(View.VISIBLE);
            fadeIn(activateDigits);
            activateDigits.setVisibility(View.VISIBLE);
        });

        activateSubmitButton.setOnClickListener(v -> {
            progress.setVisibility(View.VISIBLE);

            digits = activateDigits.getText().toString().trim();
            final CardActivationRequest request = new CardActivationRequest(digits);
            RestClient.getAuthenticated().activateCard(request).enqueue(new Callback<CardActivationResponse>() {
                @Override
                public void onResponse(Call<CardActivationResponse> call, Response<CardActivationResponse> response) {
                    CardActivationResponse cardActivationResponse = response.body();
                    if (cardActivationResponse != null && response.isSuccessful()) {
                        progress.setVisibility(View.GONE);
                        Intent activate = new Intent(ActivateMoviePassCard.this, AutoActivatedCard.class);
                        activate.putExtra(Constants.SCREENING, screeningObject);
                        activate.putExtra(Constants.SHOWTIME, selectedShowTime);
                        startActivity(activate);
                        finish();
                    } else {
                        progress.setVisibility(View.GONE);
                        try {
                            JSONObject err = null;
                            if (response.errorBody() != null) {
                                err = new JSONObject(response.errorBody().string());
                            }
                            Toast.makeText(ActivateMoviePassCard.this, err.toString(), Toast.LENGTH_SHORT).show();
                        } catch (JSONException | IOException e) {
                            e.printStackTrace();
                        }


                    }
                }

                @Override
                public void onFailure(Call<CardActivationResponse> call, Throwable t) {
                    progress.setVisibility(View.GONE);
                    Toast.makeText(ActivateMoviePassCard.this, "Server Error. Try again later", Toast.LENGTH_SHORT).show();


                }
            });
        });


    }


    @TargetApi(Build.VERSION_CODES.M)
    public void scanCard() {

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(CAMERA_PERMISSIONS, REQUEST_CAMERA_CODE);

            Intent scanIntent = new Intent(ActivateMoviePassCard.this, CardIOActivity.class);
            scanIntent.putExtra(CardIOActivity.EXTRA_UNBLUR_DIGITS, 4);
            startActivityForResult(scanIntent, Constants.CARD_SCAN_REQUEST_CODE);
        } else {
            Intent scanIntent = new Intent(ActivateMoviePassCard.this, CardIOActivity.class);
            scanIntent.putExtra(CardIOActivity.EXTRA_UNBLUR_DIGITS, 4);
            startActivityForResult(scanIntent, Constants.CARD_SCAN_REQUEST_CODE);
        }

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == Constants.CARD_SCAN_REQUEST_CODE) {
            if (data != null && data.hasExtra(CardIOActivity.EXTRA_SCAN_RESULT)) {
                final CreditCard scanResult = data.getParcelableExtra(CardIOActivity.EXTRA_SCAN_RESULT);

                activateScanCardIcon.setVisibility(View.GONE);
                activateManualInput.setVisibility(View.GONE);
                activateSubmitButton.setVisibility(View.VISIBLE);
                activateDigits.setVisibility(View.VISIBLE);
                activateDigits.setText(scanResult.getLastFourDigitsOfCardNumber());


                activateSubmitButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        digits = activateDigits.getText().toString();

                        CardActivationRequest request = new CardActivationRequest(digits);
                        RestClient.getAuthenticated().activateCard(request).enqueue(new Callback<CardActivationResponse>() {
                            @Override
                            public void onResponse(Call<CardActivationResponse> call, Response<CardActivationResponse> response) {
                                CardActivationResponse cardActivationResponse = response.body();
                                if (cardActivationResponse != null && response.isSuccessful()) {
                                    Intent intent = new Intent(ActivateMoviePassCard.this, HomeActivity.class);
                                    startActivity(intent);
                                } else {
                                    Snackbar.make(findViewById(R.id.ACTIVATE), "Incorrect card number", Snackbar.LENGTH_LONG);
                                }
                            }

                            @Override
                            public void onFailure(Call<CardActivationResponse> call, Throwable t) {
                                progress.setVisibility(View.GONE);
                                Snackbar.make(findViewById(R.id.ACTIVATE), "Server Error; Please try again. ", Snackbar.LENGTH_LONG);


                            }
                        });

                    }
                });
            }
        }
    }

    @Override
    public void onBackPressed() {
    }

    public void fadeIn(View view) {
        Animation fadeIn = new AlphaAnimation(0, 1);
        fadeIn.setInterpolator(new DecelerateInterpolator()); //add this
        fadeIn.setDuration(1000);

        AnimationSet animation = new AnimationSet(false); //change to false
        animation.addAnimation(fadeIn);
        view.setAnimation(animation);

    }

    public void fadeOut(View view) {
        Animation fadeOut = new AlphaAnimation(1, 0);
        fadeOut.setInterpolator(new DecelerateInterpolator()); //add this
        fadeOut.setDuration(1000);

        AnimationSet animation = new AnimationSet(false); //change to false
        animation.addAnimation(fadeOut);
        view.setAnimation(animation);
    }


}

File: app/src/main/java/com/mobile/activities/ActivateMoviePassCard.kt
package com.mobile.activities

import android.Manifest
import android.annotation.TargetApi
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.animation.AlphaAnimation
import android.view.animation.AnimationSet
import android.view.animation.DecelerateInterpolator
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.mobile.ApiError
import com.mobile.Constants
import com.mobile.home.HomeActivity
import com.mobile.model.Screening
import com.mobile.network.Api
import com.mobile.network.RestClient
import com.mobile.requests.CardActivationRequest
import com.moviepass.R
import io.card.payment.CardIOActivity
import io.card.payment.CreditCard
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.ac_activate_movie_pass_card.*
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import javax.inject.Inject

class ActivateMoviePassCard : AppCompatActivity() {


    @Inject
    internal lateinit var api: Api

    internal var activateCardDisposable: Disposable? = null

    internal lateinit var progress: View
    internal lateinit var activateInstructions: TextView
    internal lateinit var activateManualInput: TextView
    internal lateinit var activateDigits: EditText
    internal lateinit var activateScanCardIcon: ImageView
    internal lateinit var activateXOut: ImageView
    internal lateinit var digits: String
    var screeningObject: Screening? = null
    var selectedShowTime: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.ac_activate_movie_pass_card)


        activateManualInput = findViewById(R.id.ACTIVATECARD_MANULINPUT)
        activateDigits = findViewById(R.id.ACTIVATE_DIGITS)
        activateScanCardIcon = findViewById(R.id.ACTIVATECARD_SCAN_ICON)
        activateXOut = findViewById(R.id.ACTIVATECARD_X_OUT)
        progress = findViewById(R.id.progress)
        activateScanCardIcon.setOnClickListener { _ -> scanCard() }


        api = RestClient.getAuthenticated()

        val intent = intent ?: return
        screeningObject = intent.getParcelableExtra(Constants.SCREENING)
        selectedShowTime = getIntent().getStringExtra(Constants.SHOWTIME)


        activateXOut.setOnClickListener { _ ->
            finish()
        }

        activateManualInput.setOnClickListener { _ ->
            activatecard_instructions.text = getString(R.string.last_4)
            header1.visibility = View.GONE
            instruct1.visibility = View.GONE
            instruct2.visibility = View.GONE

            fadeOut(activateScanCardIcon)
            activateScanCardIcon.visibility = View.GONE
            fadeOut(activateManualInput)
            activateManualInput.visibility = View.GONE

            fadeIn(activateButton)
            activateButton.visibility = View.VISIBLE
            fadeIn(activateDigits)
            activateDigits.visibility = View.VISIBLE
        }

        activateButton.setOnClickListener { _: View ->
            progress.visibility = View.VISIBLE
            continueActivation()
        }
    }


    @TargetApi(Build.VERSION_CODES.M)
    fun scanCard() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(CAMERA_PERMISSIONS, REQUEST_CAMERA_CODE)
            val scanIntent = Intent(this@ActivateMoviePassCard, CardIOActivity::class.java)
            scanIntent.putExtra(CardIOActivity.EXTRA_UNBLUR_DIGITS, 4)
            startActivityForResult(scanIntent, Constants.CARD_SCAN_REQUEST_CODE)
        } else {
            val scanIntent = Intent(this@ActivateMoviePassCard, CardIOActivity::class.java)
            scanIntent.putExtra(CardIOActivity.EXTRA_UNBLUR_DIGITS, 4)
            startActivityForResult(scanIntent, Constants.CARD_SCAN_REQUEST_CODE)
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if(resultCode== Activity.RESULT_CANCELED) {
            return
        }
        if (requestCode == Constants.CARD_SCAN_REQUEST_CODE) {
            if (data != null && data.hasExtra(CardIOActivity.EXTRA_SCAN_RESULT)) {
                val scanResult = data.getParcelableExtra<CreditCard>(CardIOActivity.EXTRA_SCAN_RESULT)

                activateScanCardIcon.visibility = View.GONE
                activateManualInput.visibility = View.GONE
                activateButton.visibility = View.VISIBLE
                activateDigits.visibility = View.VISIBLE
                activateDigits.setText(scanResult.lastFourDigitsOfCardNumber)
                header1.visibility = View.GONE
                instruct1.visibility = View.GONE
                instruct2.visibility = View.GONE
                activatecard_instructions.text = getString(R.string.last_4)

                activateButton.setOnClickListener { _ -> continueActivation() }

            }
        }
    }

    private fun continueActivation() {

        digits = activateDigits.text.toString().trim { it <= ' ' }
        val request = CardActivationRequest(digits)

        activateCardDisposable?.dispose()

        activateCardDisposable = api
                .activateCardRX(request)
                .subscribe({ _ ->
                    progress.visibility = View.GONE
                    val activate = Intent(this@ActivateMoviePassCard, AutoActivatedCard::class.java)
                    activate.putExtra(Constants.SCREENING, screeningObject)
                    activate.putExtra(Constants.SHOWTIME, selectedShowTime)
                    startActivity(activate)
                    finish()

                }
                ) { error ->
                    progress.visibility = View.GONE
                    if (error is ApiError) {
                        Toast.makeText(this@ActivateMoviePassCard, error.error.message, Toast.LENGTH_SHORT).show()
                    }
                }
    }

    fun fadeIn(view: View) {
        val fadeIn = AlphaAnimation(0f, 1f)
        fadeIn.interpolator = DecelerateInterpolator() //add this
        fadeIn.duration = 1000

        val animation = AnimationSet(false) //change to false
        animation.addAnimation(fadeIn)
        view.animation = animation

    }

    fun fadeOut(view: View) {
        val fadeOut = AlphaAnimation(1f, 0f)
        fadeOut.interpolator = DecelerateInterpolator() //add this
        fadeOut.duration = 1000

        val animation = AnimationSet(false) //change to false
        animation.addAnimation(fadeOut)
        view.animation = animation
    }


    override fun onDestroy() {
        super.onDestroy()
        activateCardDisposable = null
    }

    companion object {
        private val REQUEST_CAMERA_CODE = 201
        private val CAMERA_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
    }



}
